### ISSUE TITLE
Label align with control

### ISSUE THREAD

**Issue by syaimaamri** [orbeon/orbeon-forms#2865]
Hi I would like to align the label with control as image below. Is this possible using Custom CSS classes
in Orbeon Builder 2016.1? Would appreciate your help.

![label](https://cloud.githubusercontent.com/assets/18039666/17475913/e0bc0678-5d8f-11e6-9a21-2438b9c8352d.PNG)

**Comment by avernet** [comment]
I thought we already had an RFE for this, but can't seem to be able to locate it. So I've slightly changed the title of this issue, and marked this as an RFE.

For reference, similarly, we'd like to be able to layout radio buttons and checkboxes horizontally, which is covered by #333.

**Comment by ebruchez** [comment]
Note that in general we prefer the top-aligned labels. See also [Design Better Forms  - Top align labels](https://uxdesign.cc/design-better-forms-96fadca0f49c#45de).

**Comment by ebruchez** [comment]
[+1 from user](http://discuss.orbeon.com/Orbeon-form-builder-limitation-td4661765.html)

**Comment by avernet** [comment]
[+1 from customer](https://3.basecamp.com/3600924/buckets/1921358/messages/357078080)

**Comment by avernet** [comment]
[+1 from evaluator](https://mail.google.com/mail/u/0/#inbox/15ad2be7ebee15b6)

**Comment by avernet** [comment]
[+1 from user](https://mail.google.com/mail/u/0/#inbox/FMfcgxvwxxXZKMmSCbfCzBMBfhgppBKj)

**Comment by ebruchez** [comment]
[+1 from customer](https://3.basecamp.com/3600924/buckets/7873395/messages/1576390814)

**Comment by ebruchez** [comment]
#4190 would be another way of doing it.

**Comment by ebruchez** [comment]
+1 from me for some parts of the workflow UI (#3463)

**Comment by ebruchez** [comment]
In Form Builder, we'd like this only for some controls. Could we use a label appearance?

```xml
<xf:input>
    <xf:label appearance="xxf:left">Label</xf:label>
```

This would place a class on the enclosing control, for example `xxforms-label-appearance-left`.

For something form-wide, the issue is a little different as there are alignment questions.

**Comment by ebruchez** [comment]
- [ ] generate class on control
- [ ] provide CSS to handle class

**Comment by ebruchez** [comment]
The CSS is easy with CSS grids, harder with other CSS.

**Comment by ebruchez** [comment]
One issue with form-level horizontal labels is that controls must be aligned or the result will be unacceptably ugly.

Ideally, you'd like them to be automatically sized based on content, maybe with a min/max width. But this is probably not possible right now at the level of an entire form even with bleeding-edge CSS, at least without us changing our markup, as we have nested sections and grids.

But we don't necessarily need to do this at the entire form level. Doing it at the grid level is already useful. With CSS grids and *subgrids*, you can do this for a single non-repeated grid for sure, maybe for repeated grids as well (but we need this less or not at all for repeated grids). ~~But right now, subgrids are only implemented by Firefox, and IE11 doesn't support CSS grids at all anyway.~~

Without subgrids, it's possible to do it with `display: contents` if you don't need the intermediary elements to participate to styling/layout (see ["Why we need CSS subgrid"](https://dev.to/kenbellows/why-we-need-css-subgrid-53mh). I have tried this punctually in Form Builder and it works great. But IE11 still doesn't have CSS grids.

So right now, the only way to have good support at the grid level, probably, is to have a fixed width for the labels, until we don't support IE11 anymore.

Bottom line:

- In the meanwhile, we could go with the "fixed label width" option if we want to support this at the grid or form level.
- For a single control, on the other hand, we could go ahead immediately.

UPDATE 2026-04-28: IE11 support has been removed for a long time, so ignore everything about IE11 above.

**Comment by avernet** [comment]
[+1 from user](http://discuss.orbeon.com/Dropdown-label-td4656628.html)

**Comment by avernet** [comment]
[+1 from customer](https://3.basecamp.com/3600924/buckets/18345583/messages/4652948444)

**Comment by ebruchez** [comment]
[+1 from customer](https://mail.google.com/mail/u/0/#inbox/FMfcgzQfCMqdSDljvCjPHGJLxRNFxvzq)

**Comment by ebruchez** [comment]
Relinking to #4778 for another way to do it.

**Comment by ebruchez** [comment]
Subgrids are available in browsers, but not in the PDF renderer.

**Comment by ebruchez** [comment]
I am not convinced by the subgrid solution because:

- it won't work with PDF
- but it also adds a new layout concept to Form Builder:
    - how is the label sized and aligned *within a grid cell*?

I'd like to explore the idea of having labels placed in their own grid cell:

- This would keep a single layout concept of a grid with 12 or 24 units wide.
- It would work with automatic PDF production.
- It would also allow more flexibility in positioning the labels: left, right, or anywhere within the same grid.
- It would allow aligning the label by modifying a cell setting: left, center, right, top, middle, bottom (#6706).

The drawback is that more UI work is needed in Form Builder.

- options/actions on control
    - move label to left cell (if cell is empty)
    - split cell and move label to left (if cell is large enough)
    - move separate label back to top of control
- separate label is implemented as an Explanatory Text
   - but it should indicate somehow that it is a label for a given control, not an "Explanatory Text"
   - in the markup, ideally, it could be a `<label for="">` or a `<span>` referred to by `labelledby`
- DnD/insert from toolbox
    - would simply keep label on top for now
- DdD, cut/copy/paste
    - DnD
        - would move only control or label, depending on which is moved (for now)
            - move of control would keep label in place
            - move of label would keep control in place
        - could imagine option to move both together later (with modifier key? other?)
    - cut of control
        - would leave dangling label
        - so should be prohibited?
        - or cut label alongside control
    - cut of label
        - same thing
    - copy of control/label
        - should copy control with label
    - paste
        - would ideally paste with label on left if was the way it was copied originally
        - should be smart and try to split cell, or put to left empty cell if any?
- options/actions on grid
    - put all labels to left
    - put all labels on top
- MCP server (#7701) should expose this notion
- this could be extended to hints and possibly alerts

With MCP (#7701), a lot of changes could be easily automated.

**Comment by avernet** [comment]
[+1 from user](https://mail.google.com/mail/u/0/#inbox/FMfcgzQgMCZXBgKzKtmqtVpwHsldQGKg)
